### DISCUSSION TITLE
How to approximate a nonphysical rgb colour.

### DISCUSSION THREAD

**Discussion by lishakbo** [colour-science/colour#1306] [Q&A]
I want to display a rainbow. I converted the SD colours into sRGB and then did the most obvious thing--set the negative coordinates to zero. The output is very ugly: long segments of constant hue, the yellow colour is nonexistent, weird bands. Is there a function in colour-science that produces better results? For example, gamuts on Wikipedia [look much better](https://en.wikipedia.org/wiki/File:CIE1931xy_blank.svg). I also googled photos of rainbows and prisms, they look good. What kind of function digital cameras software use?

```
def l1norm(incolor):
    return abs(incolor[0])+abs(incolor[1])+abs(incolor[2])

def rgbcutneg(incolor):
    color = incolor.copy()
    if color[0] < 0:
        color[0] = 0
    if color[1] < 0:
        color[1] = 0
    if color[2] < 0:
        color[2] = 0
    return color

rainbow_sd = []
for i in colour.SPECTRAL_SHAPE_DEFAULT.wavelengths.astype(int):
    sd_colour = colour.sd_zeros()
    sd_colour[i] = 1
    rainbow_sd.append(sd_colour)

cmfs = colour.MSDS_CMFS["CIE 1931 2 Degree Standard Observer"]
rainbow_xyz = []
rainbow_rgb = []
rainbow_rgb_positive = []
rainbow_rgb_positive_l1 = []

for i in range(len(rainbow_sd)):
    rainbow_xyz.append(colour.sd_to_XYZ(rainbow_sd[i], cmfs))
    rainbow_rgb.append(colour.XYZ_to_sRGB(rainbow_xyz[i]))
    rainbow_rgb_positive.append(rgbcutneg(rainbow_rgb[i]))
    rainbow_rgb_positive_l1.append(l1norm(rainbow_rgb_positive[i]))

maxl1 = max(rainbow_rgb_positive_l1)
rainbow_rgb_positive_normalized = []
for i in range(len(rainbow_sd)):
    rainbow_rgb_positive_normalized.append(rainbow_rgb_positive[i]/maxl1)

display = pygame.display.set_mode((500, 500))
display.fill((0,0,0))
for i in range(len(rainbow_sd)):
    pygame.draw.line(display, rainbow_rgb_positive_normalized[i]*255, (i,0), (i,99))
```
I know the sunlight has a non-constant power distribution (I will fix that later), but that couldn't be the reason it looks so bad, right?

**Comment by thomasmansencal** [comment]
Hello,

The reason why Wikipedia plot looks much better is because it has been blurred and probably desaturated. We do not have any builtin gamut mapping algorithms. A super cheap way to do this is to blend some gray:

![image](https://github.com/user-attachments/assets/1695ae3c-9c8f-40bd-bc07-c9e752bdad65)

![image](https://github.com/user-attachments/assets/4599c2fe-69e4-4d52-8bf8-411b3364a242)

**Reply by tjdcs** [reply]
Or if you want it to look like a real rainbow, blend in some sky blue (~15,000K)

**Reply by lishakbo** [reply]
Yes, thank you, your first picture looks very similar to mine.

Cutting off negative value is equivalent to finding the nearest positive sRGB colour with respect to l1 distance in rgb coordinates.
I can try Euclidean distance instead, and/or in different coordinates: XYZ, LAB. 
LAB would be computationally hard, since positive sRGB colours are not a triangular cone in these coordinates.
What do people normally do?

**Reply by lishakbo** [reply]
> Or if you want it to look like a real rainbow, blend in some sky blue (~15,000K)

Thanks, tjdcs, I will try that.

**Comment by mesvam** [comment]
One way I can think of is to render to ACES AP0 colorspace instead of sRGB, which covers all physical colors, then use an image editor for gamut mapping.
